### ISSUE TITLE
Autosubmit GUI Next Iteration Discussion

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#522]
In GitLab by @wuruchi on Apr 16, 2020, 15:33

@mcastril @kserradell @dbeltrankyl @pablohe 

In this issue, we take a look at design and architecture ideas for the next iteration of Autosubmit GUI. We also analyze and discuss design and architecture choices made in other platforms that play a similar role than what we currently plan for Autosubmit GUI.

You can also comment, suggest your own contributions, doubts, etc.

### Architecture

Starting at [Cylc-8 Architecture](https://cylc.github.io/cylc-admin/cylc-8-architecture.html#cylc-ui-server) we have that:

They adopt a Hub Proxy model inspired by `JupiterHub`, where we have a central **Hub** that launches a Web Proxy service that will handle the requests of the users. Then, when a user tries to interact with his experiment or workflow, the **Hub** spawns a **UI Server** under that user's account. In Autosubmit terms, it would be as if the **Hub** is opening a terminal logged in as the user and setting it ready to launch Autosubmit commands, but without the terminal.

We have that this model effectively allows the **Hub** to run a **UI Server** as if it were the user and in that way access the user's files, modify them, launch commands, etc. However, the actions of the **UI Server** are restricted to what we, the developers, intend. 

The **UI Server** needs to communicate with the workflow manager in some way, they (Cylc-8 team) achieve this by using [ZeroMQ](http://zguide.zeromq.org/page:all), which is a messaging library that allows for efficient communication between systems over a network. 

For our purposes, **ZeroMQ** could allow us to implement communication between our **UI Server** and the Autosubmit instance that it has to spawn to (for example) change the status of a job. The request will be launched from the *UI Server** and received by the Autosubmit CLI and then we capture the result and send it to the **UI Server**. Remember that through all these operations, the **UI Server** has made it possible to work as if it were the user. 

**Important**: So far we have defined that there is 1 **Hub** and 1 **Web Proxy**, but we have 1 **UI Server** spawned for each user that tries to use the GUI. Furthermore, this **UI Server** can access the user's files and launch commands as if it were the user.

The **Web Proxy** plays another important role, it redirects the user requests from the GUI to the user's specific URL that shows the information retrieved by the **UI Server** specific to that user.

### Comments about Autosubmit GUI

Our current implementation is centralized, there is a single API Server that works with specific requests and it cannot execute commands as if it were a specific user, only as `webadmin` (the user that is currently running the server in `bscesweb04`). There are ways to make this work though.  

On the other hand, we have this decentralized system where a **Hub** spawns **UI Servers** for each user. Currently, I do not see any big barrier that would make Autosubmit GUI not work under this model, for example, **Autosubmit API** can keep working as it is and it would be just another resource that the **UI Server** requests information from. Moreover, it seems to suit the interaction needs we have planned for the next iteration of **Autosubmit GUI**. The only significant bump in the road is to get familiar with the tools involved in the creation of this infrastructure.

### Ideas for Autosubmit

Looking at the Cylc-7 documentation, I found a couple of things that could be useful and not prohibitively hard to implement:

1. For Autosubmit GUI: In job information, we not only show the minutes it has been on the queue but when exactly it was submitted, it started, and finished. We could also include the last line of the log of that job, but this would require a change of paradigm in the logging system, which is something already in the backlog.

2. For Autosubmit: Cycl has a function **scan** that lists all the Cylc instances running on that machines. We can implement something similar so the users can easily check if some Autosubmit workflow is accidentally (or not) running on the background.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 16, 2020, 15:58

Thanks @wuruchi for the complete information. Which benefits has using a Web Proxy? I understand the UI Server as kind of a shell instance that allows to send commands as if was the user who sends them (even if it's in graphical mode and not typing anything). You say the Web Proxy does the redirection from the Hub to the UI server, but if the user is previously identified then it is not possible to do this with session or post variables?

In this scenario I understand the ZeroMQ is just a middleware to send messages in a standard format, either serialized or using objects.

> **Autosubmit API** can keep working as it is and it would be just another resource that the **UI Server** requests information from 

And in this case how would you suggest to handle the authentication or session status?

> For Autosubmit GUI: 

Yes, sometimes I have been looking in the GUI and missed this information, have real timestamps. Another twist, but very reliable with the information that we have, would be to filter all the *SIM* jobs (if there's any), get the total simulated time (sum of SIM running times avoiding job repetitions) and divide by the simulated time chunk*chunk_size*nchunks - startdate. That would give you an estimated number of SYPD. If you count queuing times you would have ASYPD. Well, we can define this a little bit more but this is something that definitely we will have one day.

> For Autosubmit: 

It can be useful indeed.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 16, 2020, 18:46

Hello @mcastril 

>Which benefits has using a Web Proxy? I understand the UI Server as kind of a shell instance that allows to send commands as if was the user who sends them (even if it's in graphical mode and not typing anything). You say the Web Proxy does the redirection from the Hub to the UI server, but if the user is previously identified then it is not possible to do this with session or post variables?

It will be precisely through the Web Proxy that we will handle authentication, session, and any session variables. Also, it will create a connection to the user's **UI Server** through WebSockets. We can say that the Web Proxy supplies a set of functionalities that would be too complex to implement individually.

>And in this case how would you suggest to handle the authentication or session status?

Now that you mention this, I remember that we currently have some experiments that can't be visualized through the GUI due to permissions. I guess the **Hub** should also start an Autosubmit API server under the user account to get full permissions.

>That would give you an estimated number of SYPD. If you count queuing times you would have ASYPD. Well, we can define this a little bit more but this is something that definitely we will have one day.

What is the meaning of SYPD?

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 17, 2020, 10:47

> Now that you mention this, I remember that we currently have some experiments that can't be visualized through the GUI due to permissions. I guess the **Hub** should also start an Autosubmit API server under the user account to get full permissions. 

So you see the architecture as having a Hub using Web Proxies and UI Server, UI Server launching the API server and also managing the requests sent by the user.

> What is the meaning of SYPD? 

Simulated years per day. It is the "standard" way of measuring throughput in Earth System modelling. As the length of the timestep depends on the configuration, this is a more objective way to compare different scales. 

There is a lot of effort going on on evaluating and comparing the SYPD for a whole project. We currently take these measures using a job template in Auto-EC-Earth. Having this metrics at the Autosubmit level is much more convenient because it can be generalized to other Auto-models and also gives another reason to use workflow managers. So workflow managers improve the ASYPD (by minimizing waits, handling errors...) and also are able to measure these numbers.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jun 5, 2020, 08:25

Hi @pablohe 

Yeah, I agree. We will change it.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jun 29, 2020, 12:54

Hello @pablohe 

You are right, it is not nice to have it at the bottom. I am looking at a way to place the tool next to the Selection Panel: 

TreeView -> Selection Tool -> Job Selection Panel

What do you think @mcastril ?

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2020, 13:07

removed milestone

**Comment by kinow** [comment]
In GitLab by @wuruchi on Aug 20, 2020, 09:23

Hi @pablohe 

The description is now at the bottom of the page:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/fa2473dce8cdb8b073a39e779d710925/image.png)

Have you any problems to visualize it?

**Comment by kinow** [comment]
In GitLab by @pablohe on Aug 20, 2020, 15:08

Yes! It is there! Thanks

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jan 13, 2021, 13:15

Hi @wuruchi ,

The issue title is ambiguous since the next iteration discussion may have happened already since this issue is 8month ago. 

This issue needs an update, or be close and be mentioned in the main GUI issue since I'm sure there are newer developments than this one, right?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 13, 2021, 13:29

closed

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jul 23, 2021, 15:32

Hello @mcastril @dbeltrankyl 

I have been reviewing some options to implement the status change from the GUI. Here are my conclusions:

* The process that changes the status of the jobs of an experiment should be launched from the API and should run under the owner of the experiment.
* I haven't found a reliable way to "impersonate" the user from the API. This means that we would require that the user provides his/her password. Then, using the username and password information we can launch a remote command through `paramiko` (aiming at `bscesautosubmit01` o `bscesautosubmit02`).
* The user might need to provide his/her password every time we need to launch a command; however, we can avoid this by requesting the password only once and then encrypt it and store it in an independent database (`sqlite3` should be enough) that has read and write permissions only for the `webadmin` user (the API user). Then, we launch the `paramiko` tasks reusing that information. 
* Every `paramiko` command should include a previous command that loads the required autosubmit library version. We can get the version from the configuration files of the experiment. @dbeltrankyl I recall you mentioned that it is possible to do this, I mean, including two consecutive commands in a single `paramiko` task. 
* We will be able to capture the output of the command from the `paramiko` object and provide a result to the user in the form of a message and/or GUI update.
* This process should also allow us to launch `autosubmit run` or `autosubmit create` commands.
* The main hurdle could be convincing our users that it is safe to provide their passwords. To facilitate this process, a `Profile` module should be implemented in the GUI where users can set their password and configure other relevant settings.

As usual, this proposal is open to suggestions.

**Comment by kinow** [comment]
In GitLab by @kserradell on Jul 23, 2021, 15:48

Thanks for sharing your conclusions, Wilmer.

Regarding the storage of the password, I would only do it during the session. As soon as the session expires, the encrypted password is deleted. Less risky regarding security and we skip problems if the user changes the password.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jul 23, 2021, 16:18

Hello @kserradell 

> I would only do it during the session. As soon as the session expires, the encrypted password is deleted.

Sure, we can implement that using a token system.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 23, 2021, 17:19

Thanks @wuruchi for the detailed summary.

Remind that the `setstatus` can be done using the update-list (always that AS is running) which should facilitate the task, reducing it to the creation of a file instead of launching a command. In any case, it's true that you need to have permissons and the best way is doing it with the experiment owner user, in case that the directory is not shared with the group.

I agree on the security concerns of storing passwords. It would be great if we could, somehow, exploit the CAS session, given that the LDAP users are the same ones as in our machines.  

If we decide to use Paramiko, one possitibility is that users store their public ssh key instead of their user/password, like we do with GitLab: https://earth.bsc.es/gitlab/-/profile/keys . If we go for this option we could automatically delete the keys, let's say every month, forcing users to update their key. In this case, how would you launch the paramiko instance from the web?

But maybe we can explore other options that using Paramiko. Could we add a feature in Autosubmit so that it raises a microservice / REST server listening for connections? An issue could be the authentication... An option could be that Autosubmit provides a token on startup, that you can copy and use from the GUI to authenticate in that particular Autosubmit instance. But I agree that this would limit the available functionality as you will not able to launch new Autosubmit instances from the GUI, for example.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 26, 2021, 16:02

Hi, as discussed now I had some kind of lapsus when I described the ssh-key solution, @wuruchi is right that we can't do it like that.

What I suggest know is that users append the API public keys inside their authorized keys file (by typing a command that the GUI will provide). This is safer than storing the passwords because passwords are also used for LDAP authentication. We should also provide the command to revoke the access whenever the user wants to do it.

We could also use additional options to increase security: https://www.ssh.com/academy/ssh/authorized_keys/openssh

For example, it seems that it's possible to restrict the access to a given user or IP.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 26, 2021, 16:31

For additional security, there are also options to restrict the commands that the authorisation is granted to, to only one (or a few of them by using a wrapper script):

https://ctrlnotes.com/restrict-a-user-to-ssh-forced-command/#

https://security.stackexchange.com/questions/118688/security-of-only-allowing-a-few-vetted-commands-using-ssh-original-command

We could create an authorized-scripts file and only include the set of commands that the API would need to execute in order to interact with the experiment.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jul 27, 2021, 15:54

Hello @mcastril 

Thank you for the information. I will test the `authorized keys` mechanism and its options as you have described. I will post in this issue with my results.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Aug 12, 2021, 14:12

Hello @mcastril @kserradell @dbeltrankyl 

I have been trying to establish a `ssh` connection from the `bscesweb04` as `webadmin` user to the workstation `bscearth318.int.bsc.es` with the following results:

- `ping` command from `bscesweb04` to the workstation fails with message `ping: sendmsg: Operation not permitted`.
- I generated a public ssh key in under user `webadmin`, copied it in the `authorized_keys` file of my `wuruchi` user folder, as it is usually done when configuring a passwordless connection. Then, the `ssh` connections fails with a timeout error.
- I tried to establish the `ssh` connection through a `python` script using the `paramiko` library but I also get a timeout error.

At this point, I suspect there is some restriction for outgoing ssh connections in the server `bscesweb04`. It makes sense as it is a production server. Is there any way to confirm this? Can this restriction be lifted?

Meanwhile, I will start to evaluate other options.

**Comment by kinow** [comment]
In GitLab by @kserradell on Aug 12, 2021, 14:46

Hello @wuruchi, yes, connections in this machine are restricted. Would it work for you to open only to `bscearth000` (instead of open all WS)?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Aug 12, 2021, 14:50

Hello @kserradell 

> Would it work for you to open only to `bscearth000` (instead of open all WS)?

Yes, that would be very helpful.

**Comment by kinow** [comment]
In GitLab by @kserradell on Aug 12, 2021, 14:53

I will ask then.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 7, 2022, 10:50

unassigned @wuruchi

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 8, 2022, 10:21

moved to autosubmitreact#21

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 9, 2022, 16:28

mentioned in issue autosubmitreact#27

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 9, 2022, 16:40

mentioned in issue autosubmitreact#30
